File: DotNet/Node.Lib/Utility/Email.cs
#region (C) enfotech & consulting, Inc. 2005
// 
// All rights are reserved. 
// 
// File:		Email.cs
// Company:		enfoTech & Consulting Inc.
// OS:			Windows XP Pro (SP1, English)
// Compiler:	Visual Studio .NET (Version 8.0.50215)
//				Microsoft .NET Framework 2.0 (Version 2.0.50215)
// History:		04/25/2005 Danwen Sun Creation
// 
#endregion 

using System;
using System.Collections.Generic;
using System.Text;
using System.Collections;
using System.Net;
using System.Net.Mail;
using System.ComponentModel;

namespace Node.Lib.Utility
{
    /// <summary>
    /// Provides properties and methods for sending message using the <see cref="System.Net.Mail">System.Net.Mail</see> namespace.
    /// </summary>
    public class Email
    {
		/// <summary>
		/// Constant value for HTML format.
		/// </summary>
		public const string MAIL_FORMAT_HTML = "HTML";
		/// <summary>
		/// Constant value for Text format.
		/// </summary>
		public const string MAIL_FORMAT_TEXT = "TEXT";
		/// <summary>
		/// Constant value for UTF7 encoding.
		/// </summary>
		public const string ENCODING_UTF7 = "UTF7";
		/// <summary>
		/// Constant value for UTF8 encoding.
		/// </summary>
		public const string ENCODING_UTF8 = "UTF8";
		/// <summary>
		/// Constant value for Ascii encoding.
		/// </summary>
		public const string ENCODING_ASCII = "ASCII";
        /// <summary>
        /// Constant value for Unicode encoding.
        /// </summary>
        public const string ENCODING_UNICODE = "UNICODE";
        
		private const string DEFAULT_PORT = "25";

        private bool isCompleted = false;
        private bool isSSL = false;

        private string errorMessage = null;
		private string host = null;
		private string port = null;
		private string sender = null;
		private string subject = null;
		private string body = null;
		private string userName = null;
		private string password = null;
		private string mailFormat = null;
		private string encoding = null;

		private ArrayList toList = null;
		private ArrayList ccList = null;
		private ArrayList bccList = null;
		private ArrayList attachments = null;

		/// <summary>
		/// Initializes a new instance of the <see cref="EAF.Lib.Utility.Email">Email</see> class.
		/// </summary>
		public Email()
		{
        }

        #region public function
        /// <summary>
        /// Sends an email.
        /// </summary>
        public void Send()
        {
            Send(false);
        }

        /// <summary>
		/// Sends an e-mail.
		/// </summary>
        /// <param name="bAsync">It is true, if send email using async method; otherwise, false.</param>
        public void Send(bool bAsync)
		{
			if (this.sender == null || this.toList == null)
				throw new Exception("There is sender or receiver information!");

			MailMessage mail = new MailMessage();

            mail.From = new MailAddress(this.sender);
			if (this.subject != null)
				mail.Subject = this.subject;
			
			if (this.body != null)
				mail.Body = this.body;

			if (this.toList != null)
			{
				foreach (object to in this.toList)
                    mail.To.Add(new MailAddress(to + ""));
			}
			if (this.ccList != null)
			{
				foreach (object cc in this.ccList)
                    mail.CC.Add(new MailAddress(cc + ""));
			}
			if (this.bccList != null)
			{
				foreach (object bcc in this.bccList)
                    mail.Bcc.Add(new MailAddress(bcc + ""));
			}
			if (this.mailFormat != null)
			{
				switch (this.mailFormat.ToUpper())
				{
					case MAIL_FORMAT_HTML: mail.IsBodyHtml = true; break;
					default: mail.IsBodyHtml = false; break;
				}
			}
			if (this.encoding != null)
			{
				switch (this.encoding.ToUpper())
				{
					case ENCODING_UTF7: 
                        mail.BodyEncoding = Encoding.UTF7;
                        mail.SubjectEncoding = Encoding.UTF7;
                        break;
					case ENCODING_UTF8: 
                        mail.BodyEncoding = Encoding.UTF8;
                        mail.SubjectEncoding = Encoding.UTF8;
                        break;
                    case ENCODING_UNICODE: 
                        mail.BodyEncoding = Encoding.Unicode;
                        mail.SubjectEncoding = Encoding.Unicode;
                        break;
					default: 
                        mail.BodyEncoding = Encoding.ASCII;
                        mail.SubjectEncoding = Encoding.ASCII;
                        break;
				}
			}
			if (this.attachments != null)
			{
                foreach (object file in attachments)
                {
                    if (file is String)
                        mail.Attachments.Add(new Attachment(file + ""));
                    else if (file is Attachment)
                        mail.Attachments.Add((Attachment)file);
                }
			}
            SmtpClient smtp = new SmtpClient();
            if (this.host != null)
                smtp.Host = this.host;
            if (this.port != null)
                smtp.Port = int.Parse(this.port);

            if (this.userName != null && this.password != null)
            {
                NetworkCredential cred = new NetworkCredential(this.userName, this.password);
                smtp.Credentials = cred;
            }
            smtp.EnableSsl = this.isSSL;
            if (bAsync)
            {
                smtp.SendCompleted += new SendCompletedEventHandler(smtp_SendCompleted);
                smtp.SendAsync(mail, "sending message");
            }
            else
            {
                try
                {
                    smtp.Send(mail);
                }
                catch (Exception ex)
                {
                    this.errorMessage = ex.ToString();
                }
                this.isCompleted = true;
            }
        }
        #endregion

        #region properties
        /// <summary>
		/// Gets or Sets a SMTP Host.
		/// </summary>
		public string Host
		{
			get { return this.host; }
			set { this.host = value; }
		}

		/// <summary>
		/// Gets or Sets a SMTP Port. Default is <b>25</b>.
		/// </summary>
		public string Port
		{
			get { return this.port; }
			set { this.port = value; }
		}

		/// <summary>
		/// Gets or Sets a sender email address.
		/// </summary>
		public string Sender
		{
			get { return this.sender; }
			set { this.sender = value; }
		}

		/// <summary>
		/// Gets or Sets a collection of to email addresses.
		/// </summary>
		public ArrayList ToList
		{
			get { return this.toList; }
			set { this.toList = value; }
		}

		/// <summary>
		/// Gets or Sets an email subject.
		/// </summary>
		public string Subject
		{
			get { return this.subject; }
			set { this.subject = value; }
		}

		/// <summary>
		/// Gets or Sets an email body content.
		/// </summary>
		public string Body
		{
			get { return this.body; }
			set { this.body = value; }
		}

		/// <summary>
		/// Gets or Sets the user name for SMTP Host, for security email sending.
		/// </summary>
		public string UserName
		{
			get { return this.userName; }
			set { this.userName = value; }
		}

		/// <summary>
		/// Gets or Sets the password for SMTP Host, for security email sending.
		/// </summary>
		public string Password
		{
			set { this.password = value; }
		}

		/// <summary>
		/// Gets or Sets an email body format.
		/// </summary>
		public string BodyFormat
		{
			get { return this.mailFormat; }
			set { this.mailFormat = value; }
		}

		/// <summary>
		/// Gets or Sets an email encoding.
		/// </summary>
		public string BodyEncoding
		{
			get { return this.encoding; }
			set { this.encoding = value; }
		}

        /// <summary>
        /// Gets the returned error message. If the message is empty string, that means sent successfully.
        /// </summary>
        public string ErrorMessage
        {
            get { return this.errorMessage + ""; }
        }

        /// <summary>
        /// Gets the indicator whether message is sent completely.
        /// </summary>
        public bool IsCompleted
        {
            get { return this.isCompleted; }
        }

        /// <summary>
        /// Gets or sets the indicator for enable SSL
        /// </summary>
        public bool IsSSL
        {
            get { return this.isSSL; }
            set { this.isSSL = value; }
        }

		/// <summary>
		/// Gets or Sets a collection of CC lists.
		/// </summary>
		public ArrayList CcList
		{
			get { return this.ccList; }
			set { this.ccList = value; }
		}

		/// <summary>
		/// Gets or Sets a collection of BCC lists.
		/// </summary>
		public ArrayList BccList
		{
			get { return this.bccList; }
			set { this.bccList = value; }
		}

		/// <summary>
		/// Gets or Sets a collection of Attachments.
		/// </summary>
		public ArrayList Attachments
		{
			get { return this.attachments; }
			set { this.attachments = value; }
        }
        #endregion

        #region private function
        private void smtp_SendCompleted(object sender, AsyncCompletedEventArgs e)
        {
            if (e.Error != null)
                this.errorMessage = e.Error.ToString();
            this.isCompleted = true;
        }
        #endregion

    }
}

File: EN Node for Java environment/Node/src/Node/API/Email.java
package Node.API;

import com.enfotech.basecomponent.utility.email.SendMail;
import java.util.ArrayList;

import Node.DB.DBManager;
import Node.DB.Interfaces.Configuration.ISystemConfiguration;

/**
 * <p>API Email</p>
 * <p>Company: enfoTech & Consulting, Inc.</p>
 * 
 * @author enfoTech
 * @version 2.0
 */

public class Email {
  private String Host = null;
  private String Port = null;
  private String UserID = null;
  private String Credential = null;

  /**
   * Constructs an Email API Class using the specified Email Host Settings
   * If userID or credential are null or the empty string, then the email will be sent with Credentials supplied to Email Server Host
   * @param host String Host Name of the Email Server (IP address or Domain Name)
   * @param port String Port Number for the Email Server (usually 25)
   * @param userID String User ID needed for authentication in order to send Email
   * @param credential String Password needed for authentication in order to send Email
   */
  public Email(String host, String port, String userID, String credential) {
    this.Host = host;
    this.Port = port;
    this.UserID = userID;
    this.Credential = credential;
  }

  /**
   * Constructs an Email API Class using the Email Served Specified in the Node Configuration
   * @param loggerName String loggerName used in case of error
   * @throws Exception
   */
  public Email (String loggerName) throws Exception
  {
    ISystemConfiguration configDB = DBManager.GetSystemConfiguration(loggerName);
    this.Host = configDB.GetEmailServerHost();
    this.Port = configDB.GetEmailServerPort();
    String uid = configDB.GetUserEmailUID();
    String pwd = configDB.GetUserEmailPWD();
    if (uid != null && !uid.trim().equals("") && pwd != null && !pwd.trim().equals(""))
    {
      this.UserID = uid;
      this.Credential = pwd;
    }
  }

  /**
   * Text Body Format
   */
  public static final String MAIL_FORMAT_TEXT = SendMail.MAIL_FORMAT_TEXT;

  /**
   * HTML Body Format
   */
  public static final String MAIL_FORMAT_HTML = SendMail.MAIL_FORMAT_HTML;

  /**
   * Low Priority
   */
  public static final String PRIORITY_LOW = SendMail.PRIORITY_LOW;

  /**
   * Normal Priority
   */
  public static final String PRIORITY_NORMAL = SendMail.PRIORITY_NORMAL;

  /**
   * High Priority
   */
  public static final String PRIORITY_HIGH = SendMail.PRIORITY_HIGH;

  /**
   * Send an Email
   * Uses the default values for Body Format and Priority (Email.MAIL_FORMAT_TEXT, Email.PRIORITY_NORMAL)
   * Sends the Email with no one on the CC List or BCC List
   * @param subject String Suject of the Email
   * @param content String Content of the Body of the Email
   * @param sender String Sender (usually an Email Address, but not required) of the Email
   * @param receiver String Receiver (should be an Email Address) of the Email
   * @return boolean true if successful, false otherwise
   */
  public boolean SendEmail (String subject, String content, String sender, String receiver)
  {
    return this.SendEmail(subject, content, sender, receiver, new ArrayList(), new ArrayList(), this.MAIL_FORMAT_TEXT, this.PRIORITY_NORMAL);
  }

  /**
   * Send an Email
   * Uses the default values for Body Format and Priority (Email.MAIL_FORMAT_TEXT, Email.PRIORITY_NORMAL)
   * @param subject String Suject of the Email
   * @param content String Content of the Body of the Email
   * @param sender String Sender (usually an Email Address, but not required) of the Email
   * @param receiver String Receiver (should be an Email Address) of the Email
   * @param ccList ArrayList An ArrayList of Strings that Comprise the Email List of the CC List
   * @param bccList ArrayList An ArrayList of Strings that Comprise the Email List of the CC List
   * @return boolean true if successful, false otherwise
   */
  public boolean SendEmail (String subject, String content, String sender, String receiver, ArrayList bccList, ArrayList ccList)
  {
    return this.SendEmail(subject, content, sender, receiver, bccList, ccList, this.MAIL_FORMAT_TEXT, this.PRIORITY_NORMAL);
  }

  /**
   * Send an Email
   * @param subject String Suject of the Email
   * @param content String Content of the Body of the Email
   * @param sender String Sender (usually an Email Address, but not required) of the Email
   * @param receiver String Receiver (should be an Email Address) of the Email
   * @param ccList ArrayList An ArrayList of Strings that Comprise the Email List of the CC List
   * @param bccList ArrayList An ArrayList of Strings that Comprise the Email List of the CC List
   * @param bodyFormat String Format of Body (Should be one Email.MAIL_FORMAT_TEXT or Email.MAIL_FORMAT_HTML)
   * @param priority String Priority of Email (Should be one of Email.PRIORITY_LOW, Email.PRIORITY_NORMAL, Email.PRIORITY_HIGH)
   * @return boolean true if successful, false otherwise
   */
  public boolean SendEmail (String subject, String content, String sender, String receiver, ArrayList ccList, ArrayList bccList, String bodyFormat, String priority)
  {
    boolean isSent = false;
    SendMail mail = new SendMail();
    mail.Host = this.Host;
    mail.Port = this.Port;
    if (this.UserID != null && !this.UserID.equals("") && this.Credential != null && !this.Credential.equals("")) {
      mail.UserName = this.UserID;
      mail.Password = this.Credential;
    }
    mail.Subject = subject;
    mail.Body = content;
    mail.BodyFormat = bodyFormat;
    mail.Priority = priority;
    mail.Sender = sender;
    mail.Receiver = receiver;
    ArrayList cc = ccList;
    if (cc == null)
      cc = new ArrayList();
    ArrayList bcc = bccList;
    if (bcc == null)
      bcc = new ArrayList();
    mail.CcList = cc;
    mail.BccList = bcc;
    try {
      mail.Send();
      isSent = true;
    } catch (Exception e) {
      e.printStackTrace();
    }
    return isSent;
  }

  /**
   * Send an Email
   * @param subject String Suject of the Email
   * @param content String Content of the Body of the Email
   * @param sender String Sender (usually an Email Address, but not required) of the Email
   * @param receiver String Receiver (should be an Email Address) of the Email
   * @param ccList ArrayList An ArrayList of Strings that Comprise the Email List of the CC List
   * @param bccList ArrayList An ArrayList of Strings that Comprise the Email List of the CC List
   * @param bodyFormat String Format of Body (Should be one Email.MAIL_FORMAT_TEXT or Email.MAIL_FORMAT_HTML)
   * @param attachmentsList ArrayList An ArrayList of attachments objects
   * @param priority String Priority of Email (Should be one of Email.PRIORITY_LOW, Email.PRIORITY_NORMAL, Email.PRIORITY_HIGH)
   * @return boolean true if successful, false otherwise
   */
  public boolean SendEmail (String subject, String content, String sender, String receiver, ArrayList ccList, ArrayList bccList, String bodyFormat,ArrayList attachmentsList, String priority)
  {
    boolean isSent = false;
    SendMail mail = new SendMail();
    mail.Host = this.Host;
    mail.Port = this.Port;
    if (this.UserID != null && !this.UserID.equals("") && this.Credential != null && !this.Credential.equals("")) {
      mail.UserName = this.UserID;
      mail.Password = this.Credential;
    }
    mail.Subject = subject;
    mail.Body = content;
    mail.BodyFormat = bodyFormat;
    mail.Priority = priority;
    mail.Sender = sender;
    mail.Receiver = receiver;
    ArrayList cc = ccList;
    if (cc == null)
      cc = new ArrayList();
    ArrayList bcc = bccList;
    if (bcc == null)
      bcc = new ArrayList();
    mail.CcList = cc;
    mail.BccList = bcc;
    if (attachmentsList == null)
    	attachmentsList = new ArrayList();
    mail.Attachments = attachmentsList;
    try {
      mail.Send();
      isSent = true;
    } catch (Exception e) {
      e.printStackTrace();
    }
    return isSent;
  }

}
